cpp/src/strings/convert/convert_urls.cu

```diff
@@ -164,50 +170,98 @@ std::unique_ptr<column> url_encode(strings_column_view const& strings,
+{
+  if (ch >= '0' && ch <= '9') return (ch - '0');
+  if (ch >= 'A' && ch <= 'F') return (ch - 'A' + 10);  // in hex A=10,B=11,...,F=15
+  if (ch >= 'a' && ch <= 'f') return (ch - 'a' + 10);  // same for lower case
+  return 0;
+}
+
+__device__ bool is_hex_digit(char ch)
+{
+  return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
+}
+
+// Functor for detecting character escape sequences in URL-encoded strings.
+// It returns true for a character index corresponding to the start of an
+// escape sequence, i.e.: '%' followed by two hexadecimal digits.
+struct url_decode_escape_detector {
+  size_type num_rows;
+  size_type const* d_offsets{};
+  char const* d_chars{};
+
+  __device__ bool operator()(size_type char_idx)
   {
-    if (ch >= '0' && ch <= '9') return (ch - '0');
-    if (ch >= 'A' && ch <= 'F') return (ch - 'A' + 10);  // in hex A=10,B=11,...,F=15
-    if (ch >= 'a' && ch <= 'f') return (ch - 'a' + 10);  // same for lower case
-    return 0;
+    size_type const* next_row_idx_ptr =
+      thrust::upper_bound(thrust::seq, d_offsets, d_offsets + num_rows, char_idx);
+    size_type end_char_idx = *next_row_idx_ptr;
+    return (char_idx + 2 < end_char_idx) && d_chars[char_idx] == '%' &&
+           is_hex_digit(d_chars[char_idx + 1]) && is_hex_digit(d_chars[char_idx + 2]);
   }
+};
 
-  // main functor method executed on each string
-  __device__ size_type operator()(size_type idx)
+// Functor for replacing character escape sequences in URL-encoded strings.
+// Each escape sequence interprets the following 2 characters as hex values to create the output
+// byte. For example, the sequence '%20' is converted into byte (0x20) which is a single space
+// character. Another example converts '%C3%A9' into 2 sequential bytes (0xc3 and 0xa9
+// respectively). Overall, 3 characters are converted into one byte whenever a '%' character
+// is encountered in the string.
+struct url_decode_char_replacer {
+  size_type chars_start_offset;  // index of first character, can be non-zero for sliced column
+  size_type num_esc_pos;
+  size_type const* d_esc_positions{};  // character index of escape sequences
+  char const* d_in_chars{};
+  char* d_out_chars{};
+
+  __device__ void operator()(size_type input_idx)
   {
-    if (d_strings.is_null(idx)) return 0;
-    string_view d_str = d_strings.element<string_view>(idx);
-    char* out_ptr = d_chars ? out_ptr = d_chars + d_offsets[idx] : nullptr;
-    size_type nbytes                  = 0;
-    const char* in_ptr                = d_str.data();
-    const char* end                   = in_ptr + d_str.size_bytes();
-    while (in_ptr < end)  // walk through each byte
-    {
-      char ch = *in_ptr++;
-      if ((ch == '%') && ((in_ptr + 1) < end)) {  // found '%', convert hex to byte
-        ch = static_cast<char>(16 * hex_char_to_byte(*in_ptr++));
-        ch += static_cast<char>(hex_char_to_byte(*in_ptr++));
+    char ch = d_in_chars[input_idx];
```

Comment:
This does not look right. The `input_idx` goes from 0 to the number of bytes in the chars column. Seems this should be adjusted with the `chars_start_offset`.